### PR TITLE
Add flexible timining feature in modern diag_manager

### LINKED ISSUE #956: need flexible output average capability
**Is your feature request related to a problem? Please describe.**
MOM6 uses FMS to manage its output frequency and time average. 
If model IC date is 03Z and we want daily averaged output, then the 1st daily averaged output will be 03Z-03Z but what we want is 03Z-24Z for the 1st day. So we believe current FMS is based on forecast hour to do the averaging, not the clock (i.e not the end of the day, or month)

**Describe the solution you'd like**
would like to see a flag that to write out files based on the clock and not just from forecast hour.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**
1. Fixes #956  Adds use_clock_average namelist to diag_manager_nml
If .true. the averaging is done based on the clock. For example, if doing daily averages and your start the simulation in day1_hour3, it will do the average between day1_hour3 to day2_hour 0. The default behavior will do the average between day1 hour3 to day2 hour3

2. Adds the capability to allow for different file frequencies for the same file. 
For example, for this diag_file yaml entry
```yaml
- file_name: flexible_timing%4yr%2mo%2dy%2hr
  freq: 1 1 1
  freq_units: hours hours hours
  time_units: hours
  unlimdim: time
  new_file_freq: 6 3 1
  new_file_freq_units: hours hours hours
  start_time: 2 1 1 0 0 0
  file_duration: 12 3 9
  file_duration_units: hours hours hours
  varlist:
  - module: ocn_mod
    var_name: var1
    reduction: average
    kind: r4
```

It will create a file every 6 hours for 12 hours
`flexible_timing_0002_01_01_01.nc` - using data from hour 0 to hour 6
`flexible_timing_0002_01_01_07.nc` - using data from hour 6 to hour 12

Then it will create a file every 3 hours for 3 hours
`flexible_timing_0002_01_01_13.nc` - using data from hour 12 to hour 15

Then it will create a file every 1 hour for 9 hours.
`flexible_timing_0002_01_01_16.nc` - using data from hour 15 to hour 16
`flexible_timing_0002_01_01_17.nc` - using data from hour 17 to hour 18
etc
etc

The filenames still need some more work.

3.  Adds code to write the time averaging relating variables (`time_bounds`, `T1`, `T2`, `DT`)
4.  Adds test to test these new features

**How Has This Been Tested?**
CI including added tests

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: diag_manager/fms_diag_file_object.F90 [hunk 1]
```diff
@@ -812,18 +844,57 @@ subroutine open_diag_file(this, time_step, file_is_opened)
...
 
+!< @brief Writes the time average variables (*_T1, *_T2, *_DT) in the netcdf file
+subroutine write_avg_time_metadata(fileobj, variable_name, dimensions, long_name, units)
+  class(FmsNetcdfFile_t), intent(inout) :: fileobj        !< The file object to write into
+  character(len=*)      , intent(in)    :: variable_name  !< The name of the time variables
+  character(len=*)      , intent(in)    :: dimensions(:)  !< The dimensions of the variable
+  character(len=*)      , intent(in)    :: long_name      !< The long_name of the variable
+  character(len=*)      , intent(in)    :: units          !< The units of the variable
+
+  !TODO harcodded double
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:861]: Is this for reproducibility?

File: diag_manager/fms_diag_file_object.F90 [hunk 2]
```diff
@@ -812,18 +844,57 @@ subroutine open_diag_file(this, time_step, file_is_opened)
...
+  !< Register the "nv" dimension that it needed for the time_bounds
+  call register_axis(fileobj, "nv", 2)
+  !TODO hardcodded double
+  call register_field(fileobj, "nv", "double", (/"nv"/))
+  call register_variable_attribute(fileobj, "nv", "long_name", "vertex number", str_len=13)
+
+  !< Register the "time_bounds" as a variable
+  dimensions(1) = "nv"
+  dimensions(2) = trim(time_var_name)
+  bounds_name = trim(time_var_name)//"_bounds"
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:888]: I wonder if we can think of something better than just repeating this

File: diag_manager/fms_diag_file_object.F90 (hunk 1 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:861]: I hardcoded the double to make things easy for me.

I **think** in the old diag manager if you compile with r8 all of the time related variables were r8 and if you compile with r4 all of the time related variables were r4.

**Review by thomas-robinson — CHANGES_REQUESTED** [review]
Combine the repetitive subroutines. 

Add a namelist check for your new variable to make sure it's running with the modern diag

**Review by thomas-robinson — APPROVED** [review]

**Comment by jiandewang** [comment]
@uramirez8707 can we use this feature using diag_table ? is yes can you provide a sample ? thanks

**Comment by uramirez8707** [comment]
@jiandewang unfortunately this feature is only going to be available for diag_table.yaml and the new diag_manager, which won't be production for at least a couple of months.

**Comment by jiandewang** [comment]
> @jiandewang unfortunately this feature is only going to be available for diag_table.yaml and the new diag_manager, which won't be production for at least a couple of months.

@uramirez8707 thanks for the heads up

**Comment by jiandewang** [comment]
@uramirez8707 can you tell me whether diag_table.yaml and the new diag_manager is available in FMS now ? I need to use diag_yaml to control the output frequency

**Comment by thomas-robinson** [comment]
> @uramirez8707 can you tell me whether diag_table.yaml and the new diag_manager is available in FMS now ? I need to use diag_yaml to control the output frequency. I checked the commit log in current FMS but can't find [b0334fc](https://github.com/NOAA-GFDL/FMS/commit/b0334fca4e92607d6e8bb68d509b17de62754edd) commit

@jiandewang The updated diag manager is _almost_ ready.  We are coding the updated reduction routines which will be the last step before testing can begin.  The reduction routines are needed so that the diag manager can be run through.  The time frame we are looking at is October/November.  When it's ready for testing, we would appreciate the help testing the flexible timing feature because we at GFDL don't have a use case and it appears you do.

**Comment by jiandewang** [comment]
@thomas-robinson thank you very much for the headup. Please let me know when it's ready. In EMC we are looking for this feature in our coupled system end-to-end global-worlkflow due to special requirement from DA group

**Comment by jiandewang** [comment]
@thomas-robinson can you tell me whether the new diag_manage and diag_table.yaml feature is in FMA main repo. ?

**Comment by thomas-robinson** [comment]
@jiandewang we are currently testing the tag 2024.01-beta5 which is on the dmUpdate branch.  If the tests all pass (which it seems like they finally will), we will be merging the code into main and releasing it at the end of this week as 2024.01.  There is likely to be a follow on patch next week or the week after.

**Comment by jiandewang** [comment]
@thomas-robinson I just started my first try but without success.
The FMS I am using is cloned from main couple of days ago:
commit 42f8506512e1b5b43982320f5b9d4ca1ca9cbebd (HEAD -> main, tag: 2024.01.01-beta1, origin/main, origin/HEAD)
Author: Alex Huth <huthalexandere@gmail.com>
Date:   Thu May 9 13:33:56 2024 -0400

for the ocean model I tested, using the following original diag_table works fine:
GOLD_SIS
1 1 1 0 0 0
#output files
 "ocean",    3,  "hours", 1, "hours", "time",
"ocean_model", "geolon",      "geolon",      "ocean", "all", .false., "none", 2
"ocean_model", "SSH",         "SSH",         "ocean","all",.false.,"none",2
"ocean_model", "SST",         "SST",         "ocean","all",.false.,"none",2

now I switched to diag_table.yaml but failed:
title: GOLD_SIS
base_date: 1 1 1 0 0 0
diag_files:
- file_name: ocean
  time_units: hours
  unlimdim: time
  is_ocean: true
  freq: 3 hours
  varlist:
  - module: ocean_model
    var_name: geolon
    reduction: none
    kind: r4
  - module: ocean_model
    var_name: SSH
    reduction: none
    kind: r4
  - module: ocean_model
    var_name: SST
    reduction: none
    kind: r4

 error message I got:
diag_manager_mod::diag_manager_init: Error parsing diag_table. Error reading the base date from the diagnostic table.

what's wrong in my yaml file ? Is there a namelist/flag to tell the system that we are using yaml style instead of traditioanl style ?

**Comment by uramirez8707** [comment]
I think you are missing setting the `use_modern_diag` namelist to .True.
https://github.com/NOAA-GFDL/FMS/blob/main/diag_manager/diag_data.F90#L391

**Comment by jiandewang** [comment]
> I think you are missing setting the `use_modern_diag` namelist to .True. https://github.com/NOAA-GFDL/FMS/blob/main/diag_manager/diag_data.F90#L391

thanks @uramirez8707. Now I got "You must compile with -Duse_yaml to use the option use_modern_diag". Let me recompile it and try again

**Comment by jiandewang** [comment]
I add "Duse_yaml" to compile FMS and that worked fine. But when I tried to compile MOM6 I got error:
parser/yaml_output_functions.c:372: undefined reference to `yaml_emitter_initialize

do I need to also add "Duse_yaml" to compile MOM6 ?

**Comment by uramirez8707** [comment]
Are you compiling FMS with cmake?

You shouldn't need to add that MOM6.

**Comment by jiandewang** [comment]
> Are you compiling FMS with cmake?
> 
> You shouldn't need to add that MOM6.

no, I followed the compiling steps from MOM6-examples. See https://github.com/NOAA-GFDL/MOM6-examples/wiki/Getting-started#compiling-the-models

**Comment by uramirez8707** [comment]
What system are you using?

You do need to install libyaml: https://github.com/yaml/libyaml
and include the -I include directory: https://github.com/NOAA-GFDL/mkmf/blob/master/templates/ncrc5-intel-classic.mk#L51 and the -L library directory and -llibyaml https://github.com/NOAA-GFDL/mkmf/blob/master/templates/ncrc5-intel-classic.mk#L146

**Comment by jiandewang** [comment]
I am using intel. Let me try to follow your information. Thanks @uramirez8707

**Pull request merged**